--- distcc-2.17/src/clinet.c~	2004-07-30 01:12:13.000000000 +0000
+++ distcc-2.17/src/clinet.c	2004-08-14 10:40:42.371505216 +0000
@@ -128,7 +128,7 @@
     int ret;
     char portname[20];
 
-    dcc_trace("connecting to %s port %d", host, port);
+    //dcc_trace("connecting to %s port %d", host, port);
     
     /* Unfortunately for us, getaddrinfo wants the port (service) as a string */
     snprintf(portname, sizeof portname, "%d", port);
